### PR TITLE
Adding a local executable plugin mechanism.

### PR DESCRIPTION
**Pull request by benvanik**
The plugin API is defined in the standalone executable_plugin.h header.
This is similar to executable_library.h and enables building compatible
plugins without relying on any other IREE code. For plugins that do,
though, support for iree_status_t integration for propagating useful
errors across the plugin boundry is supported.

A plugin manager provides lifetime management and import provider
registration such that we can have stateful import providers (JITs,
dynamically loaded libraries, etc). To keep the import feature optional
the manager sits behind the import provider interface and will only be
pulled in when the user actually creates the manager. There's definitely
room for improvement here on naming and simplification but this should
be enough to start experimenting with plugins.  Plugins are scanned in
reverse order for imports allowing for overrides of
builtin/statically-registered plugins by dynamic ones.

To start we have 3 plugin types built in: static, system-library, and
embedded-elf. Similar to the static executable loader the static plugin
allows for plugins statically linked into the application binary to be
registered for use. The other two allow for dynamic loading of plugins
at runtime.

The embedded ELF loader can be used when compiling in a standalone mode
to enable portable/hermetic plugin deployments. For ukernel-like plugins
that export specialized kernels this gives the same cross-platform
portability that our embedded ELF executables have. FatELF support has
been added to allow for portable plugins when using the standalone ELFs.
More complex software (JITs, etc) can use the system loader and do
sketchy stuff with ease (TLS, global state, threading, etc).

For easy integration in specialized runtime builds the
`IREE_HAL_EXECUTABLE_PLUGIN_REGISTRATION_FN` define is available to
hook into the initial plugin manager registration and give the hosting
application the ability to inject its own plugins. Otherwise the dynamic
registration will work either via the `--executable_plugin=` repeated
flag or explicit registration for users not going through the
driver_module registration path.

All IREE tools/bindings using the driver_module mechanism will get the
new flag and can specify plugins as file paths. If the `embedded:`
prefix is used or the file extension is `.sos` the embedded loader will
be used and otherwise the system one will be. Both of those types can
be toggled at a cmake level based on availability (bare metal won't
have the system loader, etc).

Examples:
```
--executable_plugin=some/system.dll
--executable_plugin=some/standalone.sos
--executable_plugin=embedded:some/standalone.so
```

NOTE: MSVC support for standalone executable plugins is broken
until #12801 and the sample has been disabled there.

### CODE REVIEW

File: runtime/src/iree/hal/local/elf/arch/x86_64_msvc.asm
Comment (ScottTodd) [inline runtime/src/iree/hal/local/elf/arch/x86_64_msvc.asm]: reminder to resolve this somehow and remove the `DO NOT SUBMIT` before merge

File: runtime/src/iree/hal/local/executable_loader.h
Comment (ScottTodd) [inline runtime/src/iree/hal/local/executable_loader.h:92]: We could use some system architecture diagrams for executables and loaders in the local HAL README (https://github.com/openxla/iree/tree/main/runtime/src/iree/hal/local), especially if we're expecting hosting applications or downstream projects to add their own plugins/loaders/etc.

(and diagrams on https://openxla.github.io/iree/extensions/ would be neat too)

File: runtime/src/iree/hal/local/executable_plugin.h
Comment (ScottTodd) [inline runtime/src/iree/hal/local/executable_plugin.h:133]: Did you intend for this example date code to be the date you wrote the code? 2023 not 2022 in that case :P

```suggestion
// NOTE: until we hit v1 the versioning scheme here is not set in stone.
// We may want to make this major release number, date codes (0x20230307),
// or some semantic versioning we track in whatever spec we end up having.
```

File: runtime/src/iree/hal/local/executable_plugin.h
Comment (ScottTodd) [inline runtime/src/iree/hal/local/executable_plugin.h:397]: Can you explain here (or at the top of the file) what an executable plugin is / what they can do? (I'm still a bit vague on the details even after reading the PR description and this code)

File: samples/custom_dispatch/cpu/plugin/README.md
Comment (ScottTodd) [inline samples/custom_dispatch/cpu/plugin/README.md:1]: Empty README?

File: samples/custom_dispatch/cpu/plugin/standalone_plugin.c
Comment (ScottTodd) [inline samples/custom_dispatch/cpu/plugin/standalone_plugin.c]: nit: grammar (also, this sample helps me understand how plugins fit into the broader architecture, thanks!)

```suggestion
// Demonstrates a standalone plugin exporting a single `import_mul` function.
// This models kernel libraries and custom intrinsics where there are only
// stateless functions. Standalone plugins can be compiled to portable ELFs that
```

File: samples/custom_dispatch/cpu/plugin/standalone_plugin.c
Comment (ScottTodd) [inline samples/custom_dispatch/cpu/plugin/standalone_plugin.c]: Ouch :/

File: samples/custom_dispatch/cpu/plugin/example.mlir
Comment (ScottTodd) [inline samples/custom_dispatch/cpu/plugin/example.mlir]: Should this have an attribute or use something other than `func.func`? I'd expect external / plugin-defined functions to be easier to spot (control+F to search for across a module, etc.)

File: runtime/src/iree/hal/local/executable_plugin.h
Comment (benvanik) [reply runtime/src/iree/hal/local/executable_plugin.h:133]: it's just an example :)
image

File: samples/custom_dispatch/cpu/plugin/example.mlir
Comment (benvanik) [reply samples/custom_dispatch/cpu/plugin/example.mlir]: externally defined functions are just externally defined functions - nothing special about them - lets us use upstream lowerings to func.call and such without requiring IREE-specific internals to leak (especially useful in plugins that can use just upstream dialects)

File: samples/custom_dispatch/cpu/plugin/README.md
Comment (benvanik) [reply samples/custom_dispatch/cpu/plugin/README.md:1]: I've been procrastinating on this... ugh :P

File: samples/custom_dispatch/cpu/plugin/example.mlir
Comment (ScottTodd) [inline samples/custom_dispatch/cpu/plugin/example.mlir]: LGTM once this test passes

Logs: https://github.com/openxla/iree/actions/runs/4696175348/jobs/8326087188?pr=12625#step:6:2625
```
/work/samples/custom_dispatch/cpu/plugin/example.mlir:117:9: error: 'llvm.load' op result #0 must be LLVM type with size, but got '!llvm.func<i32 (ptr<i8>, ptr<i8>, ptr<i8>)>'
        func.call @simple_mul_workgroup(%memref0, %memref1, %memref2, %dim, %tid) : (memref<?xf32>, memref<?xf32>, memref<?xf32>, index, index) -> ()
        ^
/work/samples/custom_dispatch/cpu/plugin/example.mlir:117:9: note: see current operation: %52 = "llvm.load"(%48) {ordering = 0 : i64} : (!llvm.ptr<ptr>) -> !llvm.func<i32 (ptr<i8>, ptr<i8>, ptr<i8>)>
```

**Review by ScottTodd — APPROVED** [review]

File: samples/custom_dispatch/cpu/plugin/example.mlir
Comment (benvanik) [reply samples/custom_dispatch/cpu/plugin/example.mlir]: darn looks like a merge issue - will fix!

**Comment by lundong** [comment]
@benvanik  Hi Ben, it looks that static plugins are not fully supported, unlike system-library and
embedded-elf. Could you please help add the full support (with samples if possible) for static plugins? We expect to be using it on ukernel plugins. Thanks a lot.
@hcindyl

**Comment by benvanik** [comment]
Static plugins are supported - you can either use the full plugin interface and `iree_hal_executable_plugin_manager_register_plugin` to register your plugin or the lighter-weight `iree_hal_executable_plugin_manager_register_provider` that just provides a function name -> function pointer lookup.

**Comment by lundong** [comment]
> Static plugins are supported - you can either use the full plugin interface and `iree_hal_executable_plugin_manager_register_plugin` to register your plugin or the lighter-weight `iree_hal_executable_plugin_manager_register_provider` that just provides a function name -> function pointer lookup.

@benvanik Unlike system_library_loader and embedded_elf_loader, it looks to me that static_library_loader doesn't have visibility on plugin_manager: executable_plugin_manager.h is not included in static_library_loader.c, and iree_hal_static_library_loader_create() doesn't have the plugin_manager argument. Can you please further clarify on it? Thank you very much.

**Comment by benvanik** [comment]
`iree_hal_static_library_loader_create` takes an `iree_hal_executable_import_provider_t` - you can get the provider from the plugin manager with `iree_hal_executable_plugin_manager_provider`.

Note that if you are using static library executable compilation you will want to avoid using imports if possible - the benefit of statically linking the codegen output _and_ your custom functions is that they can be made as efficient direct function calls and during compilation of your binary the linker can drop unused functions, inline functions, etc. Using static library executables and statically linked import providers/plugins is useful only for optionality (same code may be deployed statically and dynamically, etc) but otherwise it's not something that should be done unless you don't care about code size or execution efficiency.

**Pull request merged**
